Show Pop-over on Workspace  Welcome

PauloGasparSv opened the pull request:
### Details

This changes our logic for new accounts that access newDot for the first time!

If the new account is part of a workspace because it was invited then in their first signin the app should:
1. Open the workspace chat by default
2. Open the "Create Menu"/"Popover Menu" inside the workspace chat by itself showing the "Request Money" and "Add Attachment" options.


### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/270487
PROPOSAL: GH_LINK_ISSUE(COMMENT)


### Tests

This requires creating multiple accounts to test/QA!

1. Create a workspace in an existing NewDot account.
2. Invite a new user (an email that isn't registered in Expensify/NewDot) to that Workspace
    - **[QA ONLY]** Sadly you'll have to invite 1 new user for each device to QA this
    - **[DEV ONLY]** For engineering testing in dev, you can create 1 account and run the following query to reset the sate of `isFirstTimeNewExpensifyUser`:
    ```sql
    UPDATE nameValuePairs SET value = 'true' WHERE name = 'isFirstTimeNewExpensifyUser' AND accountID = <accountID>;
    ```
    <img src="https://user-images.githubusercontent.com/6564265/231563736-8799e3cb-601b-4e0b-818a-e6b180c2243a.png" width="250"/>
3. Create an account for that new invited user in NewDot
4. Make sure that in your first sign-in you are redirected to the Workspace Chat and the "Create Menu" opens by itself.
5. Make sure you can navigate off the workspace chat and into other screens.
6. Sign off that account and sign in again
7. Make sure you are not redirected to the workspace chat with the "Create Menu" opened.

- [ ] Verify that no errors appear in the JS console

### Offline tests

### QA Steps
Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/6564265/232133827-08538cc5-2b66-42e2-ba38-8d8d52757a73.mov

https://user-images.githubusercontent.com/6564265/232134463-dc36eecf-cd88-4622-8c50-20301defcd63.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/6564265/232135969-ca2d85a3-3a1a-4fb4-982f-a243bc4a15da.mov


</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/6564265/232136324-b7cbba1e-2c9f-452d-bae8-6cdc77791a3a.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/6564265/232135728-1957f824-e7ef-4518-aee7-a8eae95c2778.mov

</details>

<details>
<summary>iOS</summary>


https://user-images.githubusercontent.com/6564265/232140073-bb227505-25e8-4e89-909d-36c548ed6699.mov


</details>

<details>
<summary>Android</summary>


https://user-images.githubusercontent.com/6564265/232137137-3dff55c7-fbb4-4aec-85c2-54befef77ef9.mov



</details>

Commit:
Add optional parameter showPopoverMenu to Welcome.show

Commit:
Call welcome show method to check for welcome popover when opening reportActionCompose

Commit:
Bind showPopoverMenu to this so it updates the correct state

Commit:
Not take closed/archived workspace chats in consideration for new users

Commit:
Linting Welcome

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

Commit:
Add missing trailing comma to ReportActionCompose

luacmartins commented on src/libs/actions/Welcome.js:
```diff
@@ -118,9 +119,13 @@ function show({routes, showCreateMenu}) {
-        const workspaceChatReport = _.find(allReports, report => ReportUtils.isPolicyExpenseChat(report) && report.ownerEmail === email);
+        const workspaceChatReport = _.find(
+            allReports,
+            report => ReportUtils.isPolicyExpenseChat(report) && report.ownerEmail === email && report.statusNum !== CONST.REPORT.STATUS.CLOSED,
```
Why do we need `report.statusNum !== CONST.REPORT.STATUS.CLOSED`? Were we redirecting users to closed policy chat reports?

luacmartins replied:
Ah just saw this comment in the issue - https://github.com/Expensify/Expensify/issues/270487#issuecomment-1498038852

luacmartins replied:
All good here

luacmartins commented:
Looking good. Just left a comment.

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

Commit:
Add missing reference to CONST

Commit:
Remove extra CONST import

Commit:
Remove remaining reference to withwithNavigation

Commit:
Add imports to Welcome and withNavigation removed during merge with main

PauloGasparSv commented:
I think the issue is happening because we only update the `isFirstTimeNewExpensifyUser` NVP and send it back in Onyx in `ReconnectApp` and that call isn't made during this flow!

PauloGasparSv commented:
Both [problems](https://github.com/Expensify/App/pull/16962#issuecomment-1505827627) are coming from this branch, trying to fix here!

Commit:
Clear NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER after opening Popover menu on first sign-in

PauloGasparSv commented:
The problem was happening because `Welcome.show` is being called a bunch of times now (instead of only once like before).
Those extra calls came from `ReportActionCompose/index.js` inside `componentDidMount` where we added the `Welcome.show` call. That makes sense as that component is mounted/re-created every time we access another chat.

![image](https://user-images.githubusercontent.com/6564265/231597304-0ea564bd-8fe4-4f88-b682-0af56e463dea.png)

![image](https://user-images.githubusercontent.com/6564265/231597277-8eb2b609-11a7-45f1-8a91-3ba334d0e768.png)

So [as I first imagined]( https://github.com/Expensify/App/pull/16962#issuecomment-1505838563) the problem is related to the fact that we only update the `isFirstTimeNewExpensifyUser` NVP when `ReconnectApp` is called and we don't call that when a user first signs in and gets redirected to the workspace chat.

I fixed this by setting the local Onyx value for the `isFirstTimeNewExpensifyUser` NVP to false after using it cc @luacmartins

Commit:
Update showPopoverMenu comments

luacmartins commented:
@PauloGasparSv is this ready for review?

PauloGasparSv commented:
> @PauloGasparSv is this ready for review?

Not yet @luacmartins still adding evidences!

PauloGasparSv commented:
The behavior doesn't work in mWeb! I have no idea why though.
This is also reproducible on desktop, replicating the tests steps with a small Chrome windows break the feature too.

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

Commit:
Refactor Welcome logic to control the state of isFirstTimeNewExpensifyUser instead of receiving updates from Onyx

PauloGasparSv commented:
Latest update [here](https://expensify.slack.com/archives/C03TQ48KC/p1681424837861899?thread_ts=1681417598.780039&cid=C03TQ48KC) still trying to fix the bug.

Commit:
Only call Welcome.show at first-sign in from the correct enabled composer box

PauloGasparSv commented:
Didn't know PullerBear would be assigned here!
@chiragsalian @luacmartins @thesahindia could you please wait until I add the evidences to the P.R. tomorrow?

I'll ping everyone once the P.R. is ready.

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

PauloGasparSv commented:
Added the missing evidences, this is ready for a round of reviews @chiragsalian @thesahindia @luacmartins 
Thks for the patience!

luacmartins commented on src/libs/actions/Welcome.js:
```diff
@@ -113,9 +119,21 @@ function show({routes, showCreateMenu}) {
+
+            // If showPopoverMenu exists and returns true then it opened the Popover Menu succesfully and we can update the state of isFirstTimeNewExpensifyUser
+            // so the Welcome logic doesn't run again
+            if (showPopoverMenu && showPopoverMenu()) {
```
I think that showPopoverMenu is always true since the default `() => {}` is truthy is JS

luacmartins commented on src/libs/actions/Welcome.js:
```diff
@@ -113,9 +119,21 @@ function show({routes, showCreateMenu}) {
+            // If showPopoverMenu exists and returns true then it opened the Popover Menu succesfully and we can update the state of isFirstTimeNewExpensifyUser
+            // so the Welcome logic doesn't run again
+            if (showPopoverMenu && showPopoverMenu()) {
+                isFirstTimeNewExpensifyUser = false;
```
Aren't we already setting this to false in Onyx on line 128? Why do we need to do it again?

luacmartins commented on src/libs/actions/Welcome.js:
```diff
@@ -124,6 +142,9 @@ function show({routes, showCreateMenu}) {
         }
+
+        // Update state of isFirstTimeNewExpensifyUser so the Welcome logic doesn't run again
+        isFirstTimeNewExpensifyUser = false;
```
Shouldn't we set this in Onyx instead?

PauloGasparSv replied:
I couldn't make the feature work, specifically for mWeb/iOS/Android, if Onyx updated that key.

I discussed this further [here](https://github.com/Expensify/App/pull/16962#issuecomment-1507685866) and the solution came from this [slack thread](https://expensify.slack.com/archives/C03TQ48KC/p1681417598780039).

A brief explanation is:
In small screen devices, because of how the LNH works, some components are mounted twice which causes 3 calls to `Welcome.show` to be made.

Only the third call to `Welcome.show` can open the Popover menu. The problem is that by the time the 3rd `Welcome.show` call runs the device already synced up with Onyx and updated the state of that key to false : /

This is the best solution I could come up with so that key is still `true` by the time `Welcome.show` runs.

PauloGasparSv replied:
Not sure if that was the question @luacmartins so please lmk!

luacmartins replied:
Ah makes sense! Thanks!

luacmartins reviewed (commented):
Looking good. Just left a couple of comments.

luacmartins reviewed (commented):
Just one pending comment

luacmartins commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

Commit:
Remove unecessary showPopoverMenu check

PauloGasparSv commented:
> @PauloGasparSv, does this still require review from me? @luacmartins has already filled the checklist.

I just pulled from `main` and added a commit that shouldn't change logic, but I think it would be a good idea specially to prevent bugs that me and @luacmartins didn't catch

Not required though : )

---

Btw, I'm re-testing everything here and will ask for a re-review shortly.

luacmartins commented:
@thesahindia yes, please review!

luacmartins commented:
@thesahindia yes, you need to be part of the beta. What's the testing account you are using?

luacmartins commented:
I added that account to the beta. Are you using multiple accounts to test each platform? If so, we'll need to add those to the beta too

luacmartins commented:
Done!

luacmartins commented:
@PauloGasparSv we have conflicts

PauloGasparSv commented:
> @PauloGasparSv we have conflicts

Thks, fixing them here!

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

Commit:
Add missing imports removed in last merge with main

PauloGasparSv commented:
Re-test Web, Android and mWeb Chrome and the feature is still working as expected after the merge.

thesahindia reviewed (dismissed):
Works well!

cc: @luacmartins

chiragsalian commented on src/libs/actions/Welcome.js:
```diff
@@ -36,7 +37,11 @@ Onyx.connect({
     initWithStoredValues: false,
     callback: (val) => {
-        isFirstTimeNewExpensifyUser = val;
+        // If this is a first time new expensify user, let's only update the state of isFirstTimeNewExpensifyUser from true to false
```
```suggestion
        // If this is a first time new Expensify user, let's only update isFirstTimeNewExpensifyUser from true to false
```
having a mention of "state" is tripping me up a little bit since it's not a state variable (in a react sense). Just my opinion to make it simpler.

chiragsalian commented on src/libs/actions/Welcome.js:
```diff
@@ -36,7 +37,11 @@ Onyx.connect({
-        isFirstTimeNewExpensifyUser = val;
+        // If this is a first time new expensify user, let's only update the state of isFirstTimeNewExpensifyUser from true to false
+        // after running all Welcome related logic in Welcome.show
+        if (!isFirstTimeNewExpensifyUser) {
```
I'm a bit confused, you mention in the comment above this block updates isFirstTimeNewExpensifyUser from true to false. 

But for a new user, isFirstTimeNewExpensifyUser is first null from the top and then when backend returns true for `ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER` won't this be set from null to true?

Additionally from logic here if isFirstTimeNewExpensifyUser is true it will never set it to false. The condition here will give true only for `!false` so you can only change isFirstTimeNewExpensifyUser from false or null to true/false here.

So maybe your comment needs to be rephrased. Let me know if I'm mistaken.

PauloGasparSv replied:
I think my comment is **very poorly written** @chiragsalian so I'll explain each question and rework the comment too.

> But for a new user, isFirstTimeNewExpensifyUser is first null from the top and then when backend returns true for ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER won't this be set from null to true?

Yes, the first time this logic is reached is when we first set `isFirstTimeNewExpensifyUser` to true.

> Additionally from logic here if isFirstTimeNewExpensifyUser is true it will never set it to false. The condition here will give true only for !false so you can only change isFirstTimeNewExpensifyUser from false or null to true/false here.

Yes! It only updates the value of `isFirstTimeNewExpensifyUser` if it already was falsy.
Since we never update `isFirstTimeNewExpensifyUser` to true, we only set `isFirstTimeNewExpensifyUser` when it first loads through Onyx.

chiragsalian commented on src/libs/actions/Welcome.js:
```diff
@@ -113,9 +119,21 @@ function show({routes, showCreateMenu}) {
+            Onyx.set(ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER, false);
             Navigation.navigate(ROUTES.getReportRoute(workspaceChatReport.reportID));
+
+            // If showPopoverMenu exists and returns true then it opened the Popover Menu succesfully and we can update the state of isFirstTimeNewExpensifyUser
```
```suggestion
            // If showPopoverMenu exists and returns true then it opened the Popover Menu successfully, and we can update isFirstTimeNewExpensifyUser
```

chiragsalian commented on src/libs/actions/Welcome.js:
```diff
@@ -124,6 +142,9 @@ function show({routes, showCreateMenu}) {
             showCreateMenu();
         }
+
+        // Update state of isFirstTimeNewExpensifyUser so the Welcome logic doesn't run again
```
```suggestion
        // Update isFirstTimeNewExpensifyUser so the Welcome logic doesn't run again
```

chiragsalian reviewed (changes requested):
I think code LGTM, just left some minor clarifying comments and a suggestion to remove "state" from your comments since I initially thought it was a react state variable. So removing it makes it easier to read imo.

luacmartins commented:
@PauloGasparSv let's address Chirag's comment above and get this one merged today!

luacmartins reviewed (commented):
We also have conflicts now

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-workspace-popover-welcome

Commit:
Remove mentions of state from stateless Welcome.js component comments

Commit:
Rephrase NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER callback logic comment

Commit:
Remove last mention of state from Welcome.js comments

PauloGasparSv commented:
Fixed the conflicts with `main` and reworked the comments!

Re-testing everything here.

PauloGasparSv commented:
Re-tested and so far everything is working. I'm still adding iOS evidences but I think this is ready for reviews

<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/6564265/233170146-f4724014-5f9c-41e0-aa39-e6debfdae34f.mov

https://user-images.githubusercontent.com/6564265/233170232-d1f0e953-e0c2-4a85-b164-36f82aefe8b0.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/6564265/233170213-91152261-ae2a-40e5-96c8-93f5b8c629a7.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/6564265/233170224-98826c6f-4cfd-47ee-8572-13929bc209a5.mov


</details>

<details>


<summary>Desktop</summary>

https://user-images.githubusercontent.com/6564265/233170181-1a50c091-f5ca-4dec-b3d3-35219de59703.mov


</details>

<details>
<summary>iOS</summary>





</details>

<details>
<summary>Android</summary>
https://user-images.githubusercontent.com/6564265/233170196-44c52cfc-2376-465c-98aa-7bc83037ecb9.mov

</details>

chiragsalian reviewed (approved):
LGTM

chiragsalian commented:
Going ahead and merging since its was mostly minor changes since the latest approvals.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4759040515) to staging by https://github.com/chiragsalian in version: 1.3.3-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4768568536) to production by https://github.com/mountiny in version: 1.3.3-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
